src/pages/crm-products-page/CrmProductsPage.tsx

```diff
@@ -2,28 +2,204 @@ import { Link } from 'react-router-dom'
+  { value: ProductStatus.ACTIVATED, label: 'Активний' },
+  { value: ProductStatus.ARCHIVED, label: 'Архіований' }
+]
 
 const CrmProductsPage = () => {
+  const [status, setStatus] = useState('')
+  const [category, setCategory] = useState('')
+  const [searchQuery, setSearchQuery] = useState('')
+  const searchDebounce = useDebounce(searchQuery, 300)
+  const dispatch = useDispatch<AppDispatch>()
+  const {
+    productsForCrmTotalCount,
+    productsForCrmPageNumber,
+    productsForCrm,
+    isLoadingForCrm,
+    searchProductForCrm,
+    searchProductsForCrmTotalCount,
+    searchProductsForCrmPageNumber,
+    isLoadingForCrmSearch
+  } = useSelector((state: RootState) => state.newProduct)
+
+  const [paginationModel, setPaginationModel] = useState<GridPaginationModel>({
+    pageSize: PAGE_SIZE,
+    page: searchQuery
+      ? searchProductsForCrmPageNumber
+      : productsForCrmPageNumber
+  })
+  const mainCategories = useSelector(
+    (state: RootState) => state.allCategories.mainCategories
+  )
+  const handleCategoryChange = (event: SelectChangeEvent) => {
+    setCategory(event.target.value)
+  }
+  const handleStatusChange = (event: SelectChangeEvent) => {
+    setStatus(event.target.value)
+  }
+
+  const handlePaginationModelChange = useCallback(
+    (paginationModel: GridPaginationModel) => {
+      const categoryId = category === '' ? undefined : parseInt(category)
+      const productStatus =
+        status === '' ? undefined : (status as ProductStatus)
+      if (searchDebounce) {
+        dispatch(
+          searchProductsForCrm({
+            limit: PAGE_SIZE,
+            offset: paginationModel.page * PAGE_SIZE,
+            searchQuery: searchDebounce
+          })
+        )
+      } else {
+        dispatch(
+          fetchProductsForCrm({
+            limit: PAGE_SIZE,
+            offset: paginationModel.page * PAGE_SIZE,
+            productStatus,
+            productCategoryId: categoryId
+          })
+        )
+      }
+    },
+    [status, productsForCrmPageNumber, category, searchDebounce]
+  )
+
+  const searchTextChange = useCallback(
+    (event: ChangeEvent<HTMLInputElement>) => {
+      setSearchQuery(event.target.value)
+    },
+    []
+  )
+
+  useEffect(() => {
+    if (!searchDebounce) {
+      return
+    }
+    dispatch(
+      searchProductsForCrm({
+        limit: PAGE_SIZE,
+        offset: 0,
+        searchQuery: searchDebounce
+      })
+    )
+  }, [searchDebounce])
+
+  useEffect(() => {
+    setPaginationModel((model) => ({
+      ...model,
+      page: searchDebounce
+        ? searchProductsForCrmPageNumber
+        : productsForCrmPageNumber
+    }))
+  }, [searchDebounce, productsForCrmPageNumber, searchProductsForCrmPageNumber])
+
+  useEffect(() => {
+    const categoryId = category === '' ? undefined : parseInt(category)
+    const productStatus = status === '' ? undefined : (status as ProductStatus)
+
+    setPaginationModel((model) => ({ ...model, page: 0 }))
```

Comment:
wy do you use previous state here?

Reply:
It's good practice to update state based on previous value.